lib/rpmtag.h

```diff
@@ -430,6 +438,10 @@ typedef enum rpmSigTag_e {
     RPMSIGTAG_SHA256	= RPMTAG_SHA256HEADER,
     RPMSIGTAG_FILESIGNATURES		= RPMTAG_SIG_BASE + 18,
     RPMSIGTAG_FILESIGNATURELENGTH	= RPMTAG_SIG_BASE + 19,
+    RPMSIGTAG_VERITYSIGNATURES		= RPMTAG_SIG_BASE + 20,
+    RPMSIGTAG_VERITYSIGNATURELENGTH	= RPMTAG_SIG_BASE + 21,
+    RPMSIGTAG_VERITYSIGNATUREALGO	= RPMTAG_SIG_BASE + 22,
+    RPMSIGTAG_VERITYSIGNATUREBLKSZ	= RPMTAG_SIG_BASE + 23,
```

Comment:
RPMSIGTAG_FILESIGNATURES is not a good example on this, that's the *very* old and bad way. Define the RPMSIGTAG_* values to the corresponding RPMTAG_* values instead, eg like RPMSIGTAG_SHA256 does.

Reply:
The other thing here is: do we really need all these tags? Isn't length easily calculable from the actual signature, and does the block size actually matter for rpm? As in, can we not just decide that we use 4096 as the page size for rpm's purposes and that's it?

Reply:
Thanks for the feedback, much appreciated!

So first question, you are suggesting I move the tags to the range of RPMTAG_SIG_BASE like this - just want to make sure I got it right:
RPMTAG_SHA256HEADER		= RPMTAG_SIG_BASE+17,	/* s */

With regard to the different tags, then for the signature length, it depends on the key used and the algorithm. Are you suggesting we calculate the length of the signature from the length of the signature array and divide it by the number of entries?

Block size is a little tricky, as it depends on the file system block size, which depending on the file system is either fixed or page size. I was planning on adding option parameters to rpmsign allowing someone to specify them for the case they want to build rpms on one platform to be installed on another platform.

I was operating under the assumption that as long as the tag was an integer, it wouldn't take up too much space and not be a big deal?

Reply:
> So first question, you are suggesting I move the tags to the range of RPMTAG_SIG_BASE like this - just want to make sure I got it right:
RPMTAG_SHA256HEADER = RPMTAG_SIG_BASE+17, /* s */

Yes. So you'd have:
```
RPMTAG_VERITYSIGNATURES = RPMTAG_SIG_BASE+24, /* s */
[...]
RPMSIGTAG_VERITYSIGNATURES = RPMTAG_VERITYSIGNATURES,
```
...and the similarly for the other tags.

> With regard to the different tags, then for the signature length, it depends on the key used and the algorithm. Are you suggesting we calculate the length of the signature from the length of the signature array and divide it by the number of entries?

I'm not sure what I'm suggesting :smile: Storage size as such is not an issue, it's just that I find the length tag looking superfluous. Isn't it just "strlen() / 2" of the hex data - and once using base64, something you'll get from rpmBase64Decode(). It's not a value you need upfront, is it?

Reply:
> Block size is a little tricky, as it depends on the file system block size, which depending on the file system is either fixed or page size. I was planning on adding option parameters to rpmsign allowing someone to specify them for the case they want to build rpms on one platform to be installed on another platform.

Eek. This seems terrible from rpm POV.

I mean, rpm's get built on some system somewhere, and signed on another, and then installed on wide variety of anything that we have zero control (or interest) over, data in rpms simply cannot be filesystem specific. Or page-size specific for that matter - while arch specific packages are obviously arch specific, a huge part of rpms are noarch where no such assumptions can be made.

I get that on the kernel side the filesystem block / page size is the smallest unit that will be read in at once, but seems to me the verity block size should be hardwired to the lowest common denominator (something like 512 or 1024) and verification on page-in looks at the data at that denominator size.

Reply:
> Yes. So you'd have:
> 
> ```
> RPMTAG_VERITYSIGNATURES = RPMTAG_SIG_BASE+24, /* s */
> [...]
> RPMSIGTAG_VERITYSIGNATURES = RPMTAG_VERITYSIGNATURES,
> ```
> 
> ...and the similarly for the other tags.
> 
> > With regard to the different tags, then for the signature length, it depends on the key used and the algorithm. Are you suggesting we calculate the length of the signature from the length of the signature array and divide it by the number of entries?
> 
> I'm not sure what I'm suggesting  Storage size as such is not an issue, it's just that I find the length tag looking superfluous. Isn't it just "strlen() / 2" of the hex data - and once using base64, something you'll get from rpmBase64Decode(). It's not a value you need upfront, is it?
Oh that's an interesting point, I'll have a look at that with the base64 code. If I can derive the size from that, that would be a great win.

Reply:
> > Block size is a little tricky, as it depends on the file system block size, which depending on the file system is either fixed or page size. I was planning on adding option parameters to rpmsign allowing someone to specify them for the case they want to build rpms on one platform to be installed on another platform.
> 
> Eek. This seems terrible from rpm POV.
> 
> I mean, rpm's get built on some system somewhere, and signed on another, and then installed on wide variety of anything that we have zero control (or interest) over, data in rpms simply cannot be filesystem specific. Or page-size specific for that matter - while arch specific packages are obviously arch specific, a huge part of rpms are noarch where no such assumptions can be made.
> 
> I get that on the kernel side the filesystem block / page size is the smallest unit that will be read in at once, but seems to me the verity block size should be hardwired to the lowest common denominator (something like 512 or 1024) and verification on page-in looks at the data at that denominator size.

It's not pretty at all, I totally agree. The problem is that the kernel has to hold the Merkle tree in memory while files are opened, so the smaller the block size, the larger the memory footprint. The default should be PAGE_SIZE for public packages, but I would like to retain this option for those who package and maintain their own rpms, if that's OK with you?

Reply:
Allowing for options (eg at distro level) is not the issue, the real issue is that for this to be a generally meaningful feature for rpm at all, these signatures need to be architecture and filesystem independent. For rpm it's all the same whether the block size is big or small as long as its consistent across all architectures, we can't have essentially arch-specific signatures in a noarch package. So this seems like quite a deal-breaker to me, unless I'm misunderstanding something (which is certainly possible).

Reply:
OK I see your point. It isn't unreasonable to count on page size for this and mandate that it is up to the file system to cope if the file system block size and page size don't match. I will take out this tag.

I also have the base64 encoding/decoding working now, which allowed me to get rid of the signature length tag.